ScoutSuite/providers/aws/configs/regions_config.py

```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+from ScoutSuite.providers.base.configs.resource_config import ResourceConfig
+from ScoutSuite.providers.aws.aws_facade import AwsFacade
+
+
+class RegionsConfig(ResourceConfig):
```

Comment:
I'm not really convinced about the design of `RegionsConfig` and its "child" `LambdaServiceConfig`. It looks like you're not really taking advantages of our composite structure and inheritance principles, by having `RegionsConfig` **and** `LambdaServiceConfig` inherited from `ResourceConfig`, with `RegionsConfig` instantiated in `LambdaServiceConfig` with `LambdasConfig` passed as an argument.

Why not defining `RegionsConfig` as a base class, something like:

```python
class RegionsConfig(ResourceConfig):
    # TODO: add service as arg
    async def fetch_all(self, credentials, regions=None, partition_name='aws', targets=None):
        # TODO: Should be injected
        facade = AwsFacade()
        self['regions'] = await facade.build_region_list(service, regions, partition_name)
        self['regions_count'] = len(self['regions'])
```

and then just defining `LambdasConfig` as a child class of `RegionsConfig` (and removing `LambdaServiceConfig` which doesn't seem very useful, even though probably used in the current codebase), as following:

```python
class LambdasConfig(RegionsConfig):
    async def fetch_all(self, credentials, region, partition_name='aws', targets=None):
        super(LambdasConfig, self).fetch_all('lambda', region, partition_name) # TODO: modify fetch_all to match this signature

        # TODO: Should be injected
        facade = AwsFacade()
        for region in self['regions']:
                functions = {}
                for raw_function in facade.get_lambda_functions(region):
                    name, function = self.parse_function(raw_function)
                    functions[name] = function

        self['functions_count'] = len(functions)        
        self['functions'] = functions

    @staticmethod
    def parse_function(function):
        function['name'] = function.pop('FunctionName')
```
But we would have to deal with two things first, I believe:
- removing `LambdaServiceConfig` safely, ensuring compliance with the interface of the current codebase (mainly with the rule engine, in the end)
- modifiying `fetch_all` signature, to be able to pass a `service` argument. More than that, it seems like you don't need `credentials` and `targets` argument. I would suggest to modify the base class `ResourceConfig` to be more general:
```python
class ResourceConfig(dict):

    async def fetch_all(self, **kwargs):
        raise NotImplementedError()

    async def finalize(self):
        pass

Reply:
My idea was really trying to respect the composite structure as much as possible. I like the idea of making RegionConfig a base class. I feel like having one class handle two levels of the tree (in this case _regions_ and _functions_) is somewhat against the purpose of having a composite pattern, because you are hiding multiple layers of the structure in one node. Splitting the service/region-specific configs in one class and the instance-specific configs in another one is a sweet spot.

I'll explore a few different implementations and push the most interesting here for review.

> * modifiying `fetch_all` signature, to be able to pass a `service` argument. More than that, it seems like you don't need `credentials` and `targets` argument. I would suggest to modify the base class `ResourceConfig` to be more general

I agree, the fetch_all signature should be more general.

Reply:
> My idea was really trying to respect the composite structure as much as possible. I like the idea of making RegionConfig a base class. I feel like having one class handle two levels of the tree (in this case _regions_ and _functions_) is somewhat against the purpose of having a composite pattern, because you are hiding multiple layers of the structure in one node. Splitting the service/region-specific configs in one class and the instance-specific configs in another one is a sweet spot.
> 
> I'll explore a few different implementations and push the most interesting here for review.

I really like your new implementation! I think I finally got your point about having two separated classes, I was mainly confused by the terminology being used, I believe. Would you mind if I suggest some changes in the name of your classes, to make things clearer?

> I agree, the fetch_all signature should be more general.

Done, I just pushed a more general `fetch_all` method in `refactoring/resource_config`.

Reply:
> I think I finally got your point about having two separated classes [...]

I think too! Great teamwork ✋ 

> Would you mind if I suggest some changes in the name of your classes, to make things clearer?

On the contrary; I'm not satisfied with my classes names. I know we talked about removing the `Config` suffix. This would give us something like: `Resource`,  `Region`, `LambdaService`, `LambdaFunction`. Did you have another idea in mind?